### DISCUSSION TITLE
Data storage in the form of file input/output is not available when creating a web version.

### DISCUSSION THREAD

**Discussion by dolto** [bevyengine/bevy#8652] [Q&A] [answered]
Data storage in the form of file input/output is not available when creating a web version.

I don't know what to do if I want to save the game.
Chat GPT told me to use web storage, but I want to create a save and load function in the form of exporting and reading save files.

Please let me know what to do or if there is a plug-in involved!

**Comment by nicopap** [comment, marked as answer]
So you want the following: let the user download a save file with the current state of the game, then later be able to upload it so that the game can read it and resume from the save state.

This is a non-trivial process. Overall, for the save functionality, you would do the following:

1. Define your game state as a `struct GameSave`, and make sure you can load/save it from the current state of your bevy `World`. You might be interested in [`bevy_save`](https://lib.rs/crates/bevy_save)
2. Derive the [`serde`](https://lib.rs/crates/serde) traits on `GameSave` (`#[derive(Serialize, Deserialize)]`)
3. [Create a blob readable by the web browser](https://rustwasm.github.io/docs/wasm-bindgen/reference/arbitrary-data-with-serde.html)
4. Use [`js-sys`](https://lib.rs/crates/js-sys) crate to access the Javascript `window` API and [create a file and a save prompt](https://stackoverflow.com/questions/13405129/create-and-save-a-file-with-javascript/29576427#29576427)

I've no time to give you more details here, it's your job to implement things for your game. But hopefully this helps you move forward.

For the other way around, I suggest you search for "how to read a file in javascript" or something similar.

**Reply by dolto** [reply]
After posting this question, I continued searching for more information and arguing with gpt.
In the end I succeeded in saving and loading data to and from web storage. (hurray)

However, your answer looks very helpful.
(I'm not good at English, so I'll have to translate it properly on a computer later...)

I'm currently making a simple game, and I'm still studying, so I don't plan on making a complicated game.
(I haven't even used a collider yet!)

Thank you for your helpful answer!
